src/cryptsetup/cryptsetup.c

```diff
@@ -416,6 +431,142 @@ static int get_password(const char *vol, const char *src, usec_t until, bool acc
+        char *text=NULL, *id=NULL;
+        if (asprintf(&text, "Please enter PIN for pkcs11 key %s!", arg_pkcs11key) < 0)
+        {
+                log_oom();
+                return;
+        }
+
+        id = strjoina("cryptsetup:", cescape(arg_pkcs11key));
+        if (ask_password_auto(text, "drive-harddisk", id, "cryptsetup", until,
+                                0,
+                                &pins) < 0)
+        {
+                log_warning("Failed to query PIN!");
+                pkcs11_finalize();
+                return;
+        }
+        if (!strv_isempty(pins) && strlen(pins[0]) > 0)
+        {
+                CK_SLOT_ID_PTR p11_slots = NULL;
+                CK_ULONG p11_num_slots = 0;
+                CK_RV rv = 0;
+                CK_RV (*gfl)(CK_FUNCTION_LIST_PTR_PTR) = NULL;
+                CK_TOKEN_INFO info;
+                CK_ATTRIBUTE attrs[2];
+                CK_OBJECT_CLASS clazz = CKO_DATA;
+                CK_OBJECT_HANDLE keyobject = CK_INVALID_HANDLE;
+                CK_ULONG returncount;
+                CK_ATTRIBUTE valueattribute = { CKA_VALUE, NULL, 0 };
+
+                module = dlopen("opensc-pkcs11.so", RTLD_LAZY);
+                if (module == NULL)
+                {
+                        log_warning("Failed to load pkcs11 module");
+                        pkcs11_finalize();
+                        return;
+                }
+                gfl = (CK_RV (*)(CK_FUNCTION_LIST_PTR_PTR)) dlsym(module, "C_GetFunctionList");
+                if (gfl == NULL)
+                {
+                        log_warning("Failed to get pkcs11 function list function");
+                        pkcs11_finalize();
+                        return;
+                }
+                rv = gfl(&p11);
+                if (pkcs11_test_rv(rv, "get pkcs11 function list")) return;
+
+                rv = p11->C_Initialize(NULL);
+                if (rv == CKR_CRYPTOKI_ALREADY_INITIALIZED)
+                        log_info("Cryptoki library has already been initialized");
+                if (pkcs11_test_rv(rv, "C_Initialize")) return;
+
+                rv = p11->C_GetSlotList(CK_TRUE, NULL, &p11_num_slots);
+                if (pkcs11_test_rv(rv, "C_GetSlotList")) return;
+
+                p11_slots = calloc(p11_num_slots, sizeof(CK_SLOT_ID));
+                if (p11_slots == NULL)
+                {
+                        log_oom();
+                        pkcs11_finalize();
+                        return;
+                }
+
+                rv = p11->C_GetSlotList(CK_TRUE, p11_slots, &p11_num_slots);
+                if (pkcs11_test_rv(rv, "C_GetSlotList")) return;
+
+                rv = p11->C_OpenSession(p11_slots[0], CKF_SERIAL_SESSION,
+                                NULL, NULL, &session);
+                if (pkcs11_test_rv(rv, "C_OpenSession")) return;
+
+                rv = p11->C_GetTokenInfo(1, &info);
+                if (pkcs11_test_rv(rv, "C_GetToken")) return;
```

Comment:
Shouldn't you log in such cases? Above, you do log in case of errors, but here you don't. The general rule is to either log about *all* issues, or *none*. In the latter case, the log can be put on the caller side, but then this function would need to get a return value.